### DISCUSSION TITLE
Regarding level shifter

### DISCUSSION THREAD

**Discussion by easaw** [awawa-dev/HyperHDR#900] [General]
As per: https://github.com/awawa-dev/HyperHDR/wiki/Level-Shifter
I understand this is _absolutely_ needed.

However, I have a few questions that may also help others who are new to microelectronics, or (like me) have not come across the need for this kind of implementation.

> TL;DR should I just get a $3 rp2040? If yes, why do the Adafruit ones cost ~$20 and the off-brand $3?

`There are still questions about the rp2040 at the bottom that may be worth a look :)`

1. With the ESP32 Mini S2, which is USB C and has 5v 'out', do we still need a level shifter? (I realise GPIO is not 5v)
2. Are there any options for the SN74AHCT125N 74AHCT125 or 'level shifters' that come on a board or module?
( e.g. TXS0108E 8 Channel Logic Level Converter [AliExpress link](https://www.aliexpress.com/item/32253926452.html) )
3. I'm using a Pi 4 and as mentioned, Mini S2 - but in my case I have the 12v WS2815.
So:
- Do I need level shifter to 12v instead?
- Or, do I need a level shifter at all, as the (dual) data line is separate - and so far _'works fine'_ with WLED on an ESP32 C3/C6.

I note that there are some _kind-of_ answers:
https://github.com/awawa-dev/HyperHDR/discussions/890 "Adafruit boards with a built-in level-shifter." and
https://github.com/awawa-dev/HyperHDR/discussions/561#discussion-5177408 and https://github.com/awawa-dev/HyperHDR/discussions/561#discussioncomment-6713423 with lists of pi boards etc. that have them built in, but I'm not sure this is what I want/ need to do.

...After looking into the rp2040:
I found a comment saying the rp2040 datasheet says that the rp2040 is NOT technically 5v GPIO..?
https://hackaday.com/2023/04/05/rp2040-and-5v-logic-best-friends-this-fx9000p-confirms/#comment-6629749

Which seems to be stated here:
https://datasheets.raspberrypi.com/rp2040/rp2040-datasheet.pdf
"Power supply for digital GPIOs, nominal voltage 1.8V to 3.3V" at the bottom of page 12.

Your advice is appreciated.

Thank you!

**Comment by awawa-dev** [comment]
> I understand this is absolutely needed.

that's what the available technical documentation says, if you're lucky it can work with a 3.3V date, or win a prize buying a batch with a larger tolerance or your power supply always gives a voltage below 5v (or the drop occurs on the cables but it can also apply to the data lines) which causes 3.3V may balance on the lower limit. Only LED power supplies are usually not solid stable...

> "Adafruit boards with a built-in level-shifter." and
> ..After looking into the rp2040:
I found a comment saying the rp2040 datasheet says that the rp2040 is NOT technically 5v GPIO..?

the key words are "built-in level-shifter", these Adafruit boards are not typical rp2040 generic boards

> WS2815

Im not familiar with this model...maybe some day but probably ws2805 will go first. Check ws2815 technical specification.

**Reply by easaw** [reply]
It's _like_ WS2813 but 12v.
https://www.thesmarthomehookup.com/the-complete-guide-to-selecting-individually-addressable-led-strips/

Somehow I've had some luck with the ESP32 D1 R32 having no issues with the data, but then that is powered with a 12v barrel plug. Some issues with 5v over USB.

I'll be going back to ESP32 S2 Mini anyway, now this other issue was solved:
https://github.com/awawa-dev/HyperSPI/discussions/52

**Reply by awawa-dev** [reply]
WS2813 or WS2815? I gave you typical application. And for WS2815 minimum for a high state is 0.7VDD =>  ~8.4V (At least if we trust Chinese tech spec) so it's even worse especially for 3.3V ESP32 GPIO.
Edit: 
they define VDD as 12V in one place and 5V in the other, who knows which is correct:
http://www.normandled.com/upload/201808/WS2815%20LED%20Datasheet.pdf page 2/3

**Reply by easaw** [reply]
yes, sorry, I  meant it is similar to WS2813, as that may be more well known than the 12v WS2815.

WS2815 is 12v, but I think the 2 data lines should be 5v.
